sdk/core/System.ClientModel/src/Options/LoggingOptions.cs

```diff
@@ -0,0 +1,148 @@
+    private bool _isLoggingContentEnabled = DefaultIsLoggingContentEnabled;
+    private IList<string> _allowedHeaderNames = new List<string>(s_defaultAllowedHeaderNames);
+    private IList<string> _allowedQueryParameters = new List<string>(s_defaultAllowedQueryParameters);
+    private ILoggerFactory _loggerFactory = NullLoggerFactory.Instance;
+
+    /// <summary>
+    /// Gets or sets the implementation of <see cref="ILoggerFactory"/> to use to
+    /// create <see cref="ILogger"/> instances for logging.
+    /// </summary>
+    /// <remarks>If an ILoggerFactory is not provided, logs will be written to Event Source
+    /// instead. If an ILoggerFactory is provided, logs will be written to ILogger only and not
+    /// Event Source.</remarks>
+    /// <value>Defaults to <see cref="NullLoggerFactory"/>.</value>
+    public ILoggerFactory LoggerFactory
+    {
+        get => _loggerFactory;
+        set
+        {
+            AssertNotFrozen();
+
+            _loggerFactory = value;
+        }
+    }
+
+    /// <summary>
+    /// Gets or sets value indicating if request and response content should be logged.
+    /// </summary>
+    /// <value>Defaults to <c>false</c>.</value>
+    public bool IsLoggingContentEnabled
+    {
+        get => _isLoggingContentEnabled;
+        set
+        {
+            AssertNotFrozen();
+
+            _isLoggingContentEnabled = value;
+        }
+    }
+
+    /// <summary>
+    /// Gets or sets value indicating maximum size of content to log in bytes.
+    /// </summary>
+    /// <value>Defaults to <c>4096</c></value>
+    public int LoggedContentSizeLimit
+    {
+        get => _loggedContentSizeLimit;
+        set
+        {
+            AssertNotFrozen();
+
+            _loggedContentSizeLimit = value;
+        }
+    }
+
+    /// <summary>
+    /// Gets or sets a list of header names that are not redacted during logging.
+    /// </summary>
+    /// <value>Defaults to a list of common header names that do not
+    /// typically hold sensitive information.</value>
+    public IList<string> AllowedHeaderNames
+    {
+        get => _allowedHeaderNames;
+    }
+
+    /// <summary>
+    /// Gets or sets a list of query parameter names that are not redacted during logging.
+    /// </summary>
+    /// <value>Defaults to a list of common query parameters that do not
+    /// typically hold sensitive information.</value>
+    public IList<string> AllowedQueryParameters
+    {
+        get => _allowedQueryParameters;
+    }
+
+    /// <summary>
+    /// Freeze this instance of <see cref="LoggingOptions"/>.  After this method
+    /// has been called, any attempt to set properties on the instance or call
+    /// methods that would change its state will throw <see cref="InvalidOperationException"/>.
+    /// </summary>
+    public virtual void Freeze()
+    {
+        _frozen = true;
+        _allowedHeaderNames = new ReadOnlyCollection<string>(_allowedHeaderNames);
```

Comment:
Is this approach to preferred to using `.AsReadOnly`?  This is a curiosity question -- I don't know the answer offhand.

Reply:
AsReadOnly is on List<T> not IList<T>

Reply:
I can also cast to List and then call AsReadOnly but this way looked a little nicer to me.

Reply:
It looks like [the implementation of AsReadOnly](https://github.com/dotnet/runtime/blob/9e59acb298c20658788567e0c6f0793fe97d37f6/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs#L277-L278) does exactly what you've written, so it shouldn't make a difference.